libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2741,11 +2785,24 @@ string YulUtilFunctions::storageSetToZeroFunction(Type const& _type)
 		else if (_type.category() == Type::Category::Array)
 			return Whiskers(R"(
 				function <functionName>(slot, offset) {
+					if iszero(eq(offset, 0)) { <panic>() }
```

Comment:
Is there no way to ensure that at compile time?

Reply:
Here it is not possible actually. I'll check all the places where this is called. If we can ensure there, maybe we can remove this check.

Reply:
Yeah, I see the problem. Maybe let's keep it then and try to tune the optimizer.

Reply:
Hm, all the places where this is called it is set either from C++ variable which can be checked for equal to zero in case of ref types, or is hardcoded to `0`
Maybe it is the best to remove checks at all?

Reply:
The proper way to avoid the check here would be to have another version of this that just doesn't take an offset as argument and call that whenever the offset is statically zero - but that'd be quite some refactoring and complication in several layers above this - so I'd just keep the check and hope for the optimizer for now.

Reply:
On optimizing this out, this check should get optimized out if this function is inlined, right? There was some discussion about a step that is similar to inlining, except instead of doing a full inlining, it would specialize the function for certain arguments. Maybe that is what we need to deal with these cases.

Reply:
If one tracks this down, I think the most interesting cases fetch ``slot, offset := <indexAccess>(...)``, where the particular ``<indexAccess>`` always assigns ``offset`` to zero. So the optimization I'd imagine to be required here would be to recognize that all non-reverting code paths of a function movably assign a value to a return function that doesn't depend on the arguments and then pull that over to the call-site, if the generating expression is cheap - or something like that...

Reply:
We could have two (or more) different but related steps:
 - all non-reverting code paths of a function set a return variable to a certain constant value -> remove the return variable from the function and set it at call sites (similar to return variable elimination)
 - all calls to a function set a parameter to the same constant value: remove the parameter and assign the constant inside the function body

Reply:
So maybe we could revamp the function parameter removal steps we already have.

Reply:
> all non-reverting code paths of a function set a return variable to a certain constant value -> remove the return variable from the function and set it at call sites (similar to return variable elimination)
all calls to a function set a parameter to the same constant value: remove the parameter and assign the 

We could widen this from "constant values" to "movable value of low cost", if we're careful enough.

Reply:
I mean ideally this would even generalize to splitting any
```
function f() -> a, b {
  {
    a := ...expensive calculation...
  }
  {
    b := ...expensive calculation...
  }
}
```
Into
```
function f() -> a, b { a := f1() b := f2() }
function f1() -> a { a := ... }
function f2() -> b { b := ... }
```
which in the constant case (e.g. ``a := 42``) will have ``f1()`` and then ``f`` inlined...
I think that may be the most reasonable alternative to the "outlining" idea we once had...

Reply:
But it'll be tricky to figure out when this is possible, since in general the calculations will be interleaved, etc. So probably best to find a reasonable subset of cases to start - but maybe more general than just constants :-).

Reply:
If we follow that path, this would be headed towards the old optimizer reconstructing values based on data flow analysis :-). But that's not necessarily a bad thing IMO.